docs: add KeptnAppContext ref page; update KeptnApp ref page (#2894)
keptn/lifecycle-toolkit: docs/docs/reference/crd-reference/appcontext.md

---
comments: true
---

# KeptnAppContext

(Need to add description)

It works with the corresponding
[KeptnApp](app.md) resource
that has the same name and is located in the same namespace.

`KeptnAppContext` lists the tasks and evaluations
to be executed pre/post-deployment.
Tasks referenced by `KeptnAppContext` are defined in a
[KeptnTaskDefinition](taskdefinition.md)
resource.
`KeptnAppContext` identifies the task by the value of the `metadata.name` field
and does not need to understand what runner is used to define the task.
Similarly, evaluations referenced are defined in a
[KeptnEvaluationDefinition](evaluationdefinition.md)
resource and identified by the value of the `metadata.name` field;
`KeptnAppContext` does not need to understand
the data source or query being used for the evaluation.

## Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v1beta1
kind: KeptnAppContext
metadata:
  name: <app-name>
  namespace: <app-namespace>

> spec:

Reviewer: We are missing metadata and spanLinks here

Author: Fixed

  preDeploymentTasks:
  - <list of tasks>
  postDeploymentTasks:
  - <list of tasks>
  preDeploymentEvaluations:
  - <list of evaluations>
  postDeploymentEvaluations:
  - <list of evaluations>
  ...
```

## Fields

* **apiVersion** -- API version being used
   Must be set to `v1beta1`
* **kind** -- Resource type
   Must be set to `KeptnAppContext`
* **metadata**
  * **name** -- Unique name of this `KeptnAppContext` resource.
       Names must comply with the
       [Kubernetes Object Names and IDs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names)
       specification
       and match the `name` given to the associated `KeptnApp` resource..
  * **namespace** -- Namespace of this application.
       This must match the `namespace` given to
       the associated `KeptnAp` resource.

The remaining fields are required only when implementing
the release lifecycle management feature.
If used, these fields must be populated manually:

> * **spec**

Reviewer: also here, missing metadata and spanlinks

Author: fixed

---
comments: true
---

# KeptnAppContext

The `KeptnAppContext` custom resource
allows you to add metadata and links to traces for a specific application.
This enables you to enrich your traces with additional information,
making it easier to understand and analyze

> the performance of your applications.

Author: Whoops!  fixed

It works with the corresponding
[KeptnApp](app.md) resource
that has the same name and is located in the same namespace.

## Synopsis

```yaml
apiVersion: lifecycle.keptn.sh/v1beta1
kind: KeptnAppContext
metadata:
  name: <app-name>
  namespace: <app-namespace>
spec:
  metadata:
    <custom-attributes>

>     spanLinks: "<link>

Reviewer 2: the `spanLinks` field is **next to** the `metadata` field, not inside of it. And it's a list of links.
```suggestion
  spanLinks:
  - "<list of links>"
```

  preDeploymentTasks:
  - <list of tasks>
  postDeploymentTasks:
  - <list of tasks>
  preDeploymentEvaluations:
  - <list of evaluations>
  postDeploymentEvaluations:
  - <list of evaluations>

>   ...

Reviewer 2: there's nothing more 😄 
```suggestion
```

```

## Fields

* **apiVersion** -- API version being used

>    Must be set to `v1beta1`

Reviewer 2: ```suggestion
   Must be set to `lifecycle.keptn.sh/v1beta1`
```

* **kind** -- Resource type
   Must be set to `KeptnAppContext`
* **metadata**
  * **name** -- Unique name of this `KeptnAppContext` resource.
       Names must comply with the
       [Kubernetes Object Names and IDs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names)
       specification

>        and match the `name` given to the associated `KeptnApp` resource..

Reviewer 2: ```suggestion
       and match the `name` given to the associated `KeptnApp` resource.
```

  * **namespace** -- Namespace of this application.
       This must match the `namespace` given to

>        the associated `KeptnAp` resource.

Reviewer 2: ```suggestion
       the associated `KeptnApp` resource.
```

* **spec**

>   * **custom-attributes** -- list of key-value pairs

Reviewer 2: the field is called `metadata`
```suggestion
  * **metadata** -- list of key-value pairs
```

       that are propagated to the application trace as attributes.

>        For example, the following lines adds the `commit-id`

Reviewer 2: ```suggestion
       For example, the following lines add the `commit-id`
```

       and `author` information to the `KEPTN_CONTEXT`
       of the workload or `KeptnApp` where it is specified:

>        ```yaml
>        spec:
>           metadata:
>              commit-id: "1234"
>              author: "myUser"
>        ```

Reviewer 2: ```suggestion
       ```yaml
       spec:
         metadata:
           commit-id: "1234"
           author: "myUser"
       ```
```

>   * **spanLinks** -- OpenTelemetry span link
>     that connects multiple traces.

Reviewer 2: ```suggestion
  * **spanLinks** -- List of OpenTelemetry span links
    that connect multiple traces.
```

    For example, this can be used to connect
    deployments of the same application
    through different stages.
    You can retrieve the value to use
    from the JSON representation of the trace in Jaeger.
    The structure of this is:

    ```yaml
    00-<trace-id>-<span-id>-01
    ```

    After you add this field to your `KeptnAppContext` manifest,

>     you must increment the `version` number

Reviewer 2: ```suggestion
    you must increment the `version` number of the corresponding `KeptnApp`
```

       specification
       and match the `name` given to the associated `KeptnApp` resource.
     - **namespace** -- Namespace of this application.
       This must match the `namespace` given to
       the associated `KeptnApp` resource.
- **spec**
     - **metadata** -- list of key-value pairs
       that are propagated to the application trace as attributes.
       For example, the following lines add the `commit-id`
       and `author` information to the `KEPTN_CONTEXT`
       of the workload or `KeptnApp` where it is specified:

          ```yaml
          spec:
            metadata:
              commit-id: "1234"
              author: "myUser"
          ```

     - **spanLinks** -- List of OpenTelemetry span links
    that connect multiple traces.
    For example, this can be used to connect
    deployments of the same application
    through different stages.
    You can retrieve the value to use
    from the JSON representation of the trace in Jaeger.
    The structure of this is:

          ```yaml
          00-<trace-id>-<span-id>-01
          ```

          After you add this field to your `KeptnAppContext` manifest,

>           you must increment the `version` number

Reviewer 3: ```suggestion
          you must increment the `revision` number
```

# KeptnAppContext

The `KeptnAppContext` custom resource works with the corresponding
[KeptnApp](app.md) resource
that has the same name and is located in the same namespace.
It allows you to

- Add metadata and links to traces for a specific application.

>   This enables enriches your traces with additional information,

Reviewer 2: ```suggestion
  This enables you to enrich your traces with additional information that
```
